pkg/scheduler/server/internal/pool.go

```diff
@@ -0,0 +1,212 @@
+/*
+Copyright 2024 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package internal
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/rand"
+	"sync"
+	"sync/atomic"
+
+	"google.golang.org/protobuf/types/known/anypb"
+
+	schedulerv1pb "github.com/dapr/dapr/pkg/proto/scheduler/v1"
+	"github.com/dapr/kit/logger"
+)
+
+var log = logger.NewLogger("dapr.runtime.scheduler")
+
+// Pool represents a connection pool for namespace/appID separation of sidecars to schedulers.
+type Pool struct {
+	nsPool map[string]*namespacedPool
+
+	lock    sync.RWMutex
+	wg      sync.WaitGroup
+	closeCh chan struct{}
+	running atomic.Bool
+}
+
+type namespacedPool struct {
+	idx       atomic.Uint64
+	appID     map[string][]uint64
+	actorType map[string][]uint64
+	conns     map[uint64]*conn
+}
+
+// JobEvent is a triggered job event.
+type JobEvent struct {
+	Name     string
+	Data     *anypb.Any
+	Metadata *schedulerv1pb.JobMetadata
+}
+
+func NewPool() *Pool {
+	return &Pool{
+		nsPool:  make(map[string]*namespacedPool),
+		closeCh: make(chan struct{}),
+	}
+}
+
+func (p *Pool) Run(ctx context.Context) error {
+	if !p.running.CompareAndSwap(false, true) {
+		return errors.New("pool is already running")
+	}
+
+	<-ctx.Done()
+	close(p.closeCh)
+	p.wg.Wait()
+
+	return nil
+}
+
+// Add adds a connection to the pool for a given namespace/appID.
+func (p *Pool) Add(req *schedulerv1pb.WatchJobsRequestInitial, stream schedulerv1pb.Scheduler_WatchJobsServer) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	nsPool, ok := p.nsPool[req.GetNamespace()]
+	if !ok {
+		nsPool = &namespacedPool{
+			appID:     make(map[string][]uint64),
+			actorType: make(map[string][]uint64),
+			conns:     make(map[uint64]*conn),
+		}
+
+		p.nsPool[req.GetNamespace()] = nsPool
+	}
+
+	ok = true
+	var uuid uint64
+	for ok {
+		uuid = rand.Uint64() //nolint:gosec
```

Comment:
Should listen to the linter :) math/rand doesn't produce good random numbers. Use crypto/rand instead.
(But this should probably use a true UUID which has ~128 bits of entropy instead of 64...)

Reply:
This is used as indexing rather than a uuid. I can update the name to just be id to avoid confusion. Since there are no security implication here and it is used for indexing purposes, I can change this to id++.

Reply:
Ok `id++` is fine.

It's not just about security, it's also that `math/rand` doesn't guarantee good randomness, so conflicts can arise. If you can use an increment, that's better